### PR TITLE
feat: add optimal pipeline setting post

### LINKED ISSUE #375: How to setup efficient pipeline triggers
Idea for a blogpost. We should only post it if it is cool enough.

More title ideas:
- our efficient gitlab pipeline setup

Our setup in gitlab:
- in MRs
  - deploy Dev & Test is possible immediately (needs: [])
  - run build and test jobs together (more pipeline stages are not better)
  - use install caches and build artifacts, with fallback caches to master in case no package was installed
  - finetune the job triggers (needs: job: install, optional: true)
- in main / master
  - deploy to prod is also possible immediately (a merged MR from a branch will anyway have build & test succeeding, so it is safe to deploy immediately in 99% of the cases)
- review apps for care-free test deployments (link to other blog post)

### PR DESCRIPTION
**Pull request by flyck**
Closes #375

My dalle picture sucks (cancelled the midjourney subscription) but thats the draft:
image

### CODE REVIEW

File: data/blog/optimal-pipeline-setup.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+---
+publishDate: "Jan 21 2024"
+title: "How to setup efficient Pipeline Triggers"
+description: "How we tune our CI/CD pipelines for optimal workflows and minimal waiting times"
+image: "~/assets/images/thumbnails/optimal_pipeline_setup_dalle.png"
```
Comment (flyck) [inline data/blog/optimal-pipeline-setup.md:5]: need a better picture

Comment (hendriknielaender) [reply data/blog/optimal-pipeline-setup.md:5]: v1

Comment (flyck) [reply data/blog/optimal-pipeline-setup.md]: Got some more suggestions..
image
image

File: data/blog/optimal-pipeline-setup.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
+---
+publishDate: "Jan 28 2024"
+title: "How to setup efficient Pipeline Triggers"
```
Comment (hendriknielaender) [inline data/blog/optimal-pipeline-setup.md:3]: SEO:
```suggestion
title: ""Efficient CI/CD Pipeline Triggers: A Step-by-Step Guide for GitLab"
```

File: data/blog/optimal-pipeline-setup.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+description: "How we tune our CI/CD pipelines for optimal workflows and minimal waiting times"
```
Comment (hendriknielaender) [inline data/blog/optimal-pipeline-setup.md:4]: SEO: Provide more descriptive and engaging content to entice clicks.
```suggestion
description: "Discover strategies to fine-tune your CI/CD pipelines in GitLab for maximum efficiency and reduced wait times, complete with practical examples and tips."
```

File: data/blog/optimal-pipeline-setup.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+image: "~/assets/images/thumbnails/optimal_pipeline_dalle.jpg"
+imageCreditUrl: https://labs.openai.com
+tags: [cloud, infrastructure, IaC, gitlab]
```
Comment (hendriknielaender) [inline data/blog/optimal-pipeline-setup.md:7]:
```suggestion
tags: [cloud, infrastructure, IaC, GitLab, CI/CD, DevOps, pipeline, automation]
```

File: data/blog/optimal-pipeline-setup.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    - job: install
+      optional: true
+```
+
+## Deployment Job Triggers
+
+The deployment to prod on master generally should be a manual action that can be triggered
+immediately. It doesn't have to wait for the dev and test deployment to succeed again first. A
+merged MR will anyway have had all necessary safety checks succeeding in the Pull Request
+already. This make it safe to deploy immediately to prod after merge in 99% of the cases. This
```
Comment (hendriknielaender) [inline data/blog/optimal-pipeline-setup.md:52]:
```suggestion
already. This make it safe to deploy immediately to production after merge in 99% of the cases. This
```

File: data/blog/optimal-pipeline-setup.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+techniques with concrete examples for gitlab, but similar concepts could also be applied to other
+CI/CD platforms.
+
+## Caching & Artifacts
+
+Many pipeline systems have the concept of caching and artifacts, and so does
+[gitlab](https://docs.gitlab.com/ee/ci/caching/). Our recommendation is to use caches for install
+dependencies and artifacts for build results, with fallback caches to master in case the
+dependencies didn't change. This allows pull requests to skip the install job in case no dependency
+has changed. Similarly the artifact has to be build only once.
```
Comment (hendriknielaender) [inline data/blog/optimal-pipeline-setup.md:22]:
```suggestion
has changed. Similarly, the artifact has to be build only once.
```

File: data/blog/optimal-pipeline-setup.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+is one that adapts to changes in the repository, to be always as fast as possible.
+
+## The Build Container
+
+If you use big dependencies during deployment, it can be worth it to bundle them already into the
+build container. Maintaining 1 or 2 build containers is usually within reason, while it should of
+course not go out of hand.
+
+If all you need is the aws cli, note that there is an official docker images for that by AWS
+[link](https://hub.docker.com/r/amazon/aws-cli). Using a combination of such tool-specific
```
Comment (hendriknielaender) [inline data/blog/optimal-pipeline-setup.md:94]:
```suggestion
[AWS CLI Official Docker Image](https://hub.docker.com/r/amazon/aws-cli). Using a combination of such tool-specific
```

File: data/blog/optimal-pipeline-setup.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+techniques, which we also found useful in other repos. In this post we will share these general
+techniques with concrete examples for gitlab, but similar concepts could also be applied to other
+CI/CD platforms.
+
+## Caching & Artifacts
+
+Many pipeline systems have the concept of caching and artifacts, and so does
+[gitlab](https://docs.gitlab.com/ee/ci/caching/). Our recommendation is to use caches for install
+dependencies and artifacts for build results, with fallback caches to master in case the
+dependencies didn't change. This allows pull requests to skip the install job in case no dependency
```
Comment (hendriknielaender) [inline data/blog/optimal-pipeline-setup.md:21]:
```suggestion
dependencies didn't change. "This setup enables pull requests to bypass the installation job when there are no changes in dependencies."
```

File: data/blog/optimal-pipeline-setup.md (hunk 1 above)
Comment (flyck) [reply data/blog/optimal-pipeline-setup.md]: Thanks for adding these extra picture, I think it was worth it 💪🏻

**Pull request merged**